### ISSUE TITLE
Integrate source code formatter (linter) into IDE

### ISSUE THREAD

**Issue by alexose** [openscad/openscad#3565]
There's a wide variety of styles for writing OpenSCAD code:  Tabs, spaces, spaces before parenthesis, brackets, no brackets, etc.  It would be helpful to add a code formatter (a.k.a. linter) into the IDE in order to standardize the way different projects look.

There's already an effort underway here https://github.com/Maxattax97/openscad-format that might be helpful.

**Comment by t-paul** [comment]
Marking as feature request, but external calls are not likely to be accepted. Formatting should be based on the internal AST.

**Comment by alexose** [comment]
A style guide (ala [AirBnB's javascript style guide](https://github.com/airbnb/javascript)) might be a useful piece of this as well.

**Comment by alexose** [comment]
I came here to suggest this feature, but then I realized that I had already suggested it in 2021.  🤦

**Comment by jordanbrown0** [comment]
Note that [linters](https://en.wikipedia.org/wiki/Lint_(software)) are usually *not* style checkers.  Instead, they look for subtle potential errors, patterns where the program passes the compiler but might still have errors.  Variables assigned but never used, used but never assigned, constant expressions in conditionals, type mismatches, that kind of thing.

**Comment by jordanbrown0** [comment]
Having a general configurable "feed this selection through an external program" mechanism doesn't seem silly.

**Comment by jordanbrown0** [comment]
Duplicate of #2011.

**Comment by mkatychev** [comment]
Hi all, just wanted to chime in that  I've written an OpenSCAD formatter that is now ready for general use:
https://github.com/tweag/topiary/pull/845

* You can see the [before](https://github.com/tweag/topiary/blob/7afe496065b18c12dff8631e8e6a5e94979e19f4/topiary-cli/tests/samples/input/openscad.scad) and [after](https://github.com/tweag/topiary/blob/7afe496065b18c12dff8631e8e6a5e94979e19f4/topiary-cli/tests/samples/expected/openscad.scad) formatting pass in the pull request linked above.

* More details on setup and how the parser/lexer is implemented:
  https://lists.openscad.org/empathy/thread/Z4PXFLQM53WXRCYUUKKNOMHJFT4CUMEW
  installation should be simplified once the openscad feature is merged into main

* To address https://github.com/openscad/openscad/issues/3565#issuecomment-753021465: 
  though this formatter uses an external AST, the AST/CST itself is explicitly built for OpenSCAD and covers a [myriad of edge cases](https://github.com/mkatychev/tree-sitter-openscad/blob/master/test/corpus/literals.txt)

P.S.

Curious if anyone could chime in on the indentation issue brought up [here](https://github.com/tweag/topiary/pull/845#discussion_r1932656458), OpenSCAD doesn't seem to have a style guide so I went off of what I saw in the wiki book and popular libraries.

**Comment by jordanbrown0** [comment]
I don't know why I didn't close this in 2023 when I noted that it was a duplicate of #2011.
